src/Microsoft.Data.SqlClient/tests/ManualTests/SQL/SqlBulkCopyTest/TestBulkCopyWithUTF8.cs

```diff
@@ -0,0 +1,82 @@
+        {
+            string columnDefinition = "(str_col varchar(max) COLLATE Latin1_General_100_CS_AS_KS_WS_SC_UTF8)";
+            DataTestUtility.CreateTable(connection, sourceTable, columnDefinition);
+            DataTestUtility.CreateTable(connection, destinationTable, columnDefinition);
+
+            using SqlCommand insertCommand = connection.CreateCommand();
+            insertCommand.CommandText = insertQuery;
+            Helpers.TryExecute(insertCommand, insertQuery);
+        }
+
+        [ConditionalFact(typeof(DataTestUtility), nameof(DataTestUtility.AreConnStringsSetup), nameof(DataTestUtility.IsNotAzureServer), nameof(DataTestUtility.IsNotAzureSynapse))]
+        public void BulkCopy_Utf8Data_ShouldMatchSource()
+        {
+            string sourceTable = DataTestUtility.GetUniqueName("SrcUtf8DataTable");
+            string destinationTable = DataTestUtility.GetUniqueName("DstUtf8DataTable");
+            string insertQuery = $"INSERT INTO {sourceTable} VALUES('test')";
+
+            using SqlConnection sourceConnection = new SqlConnection(_connectionString);
+            sourceConnection.Open();
+            SetupTables(sourceConnection, sourceTable, destinationTable, insertQuery);
+
+            using SqlCommand countCommand = new SqlCommand($"SELECT COUNT(*) FROM {destinationTable}", sourceConnection);
+            long initialCount = Convert.ToInt64(countCommand.ExecuteScalar());
+
+            using SqlCommand sourceDataCommand = new SqlCommand($"SELECT str_col FROM {sourceTable}", sourceConnection);
+            using SqlDataReader reader = sourceDataCommand.ExecuteReader(CommandBehavior.SequentialAccess);
+
+            using SqlConnection destinationConnection = new SqlConnection(_connectionString);
+            destinationConnection.Open();
+
+            using SqlBulkCopy bulkCopy = new SqlBulkCopy(destinationConnection)
+            {
+                EnableStreaming = true,
+                DestinationTableName = destinationTable
+            };
+
+            try
+            {
+                bulkCopy.WriteToServer(reader);
+            }
+            catch (Exception ex)
+            {
+                Assert.Fail($"Bulk copy failed: {ex.Message}");
+            }
+
+            long finalCount = Convert.ToInt64(countCommand.ExecuteScalar());
+            Assert.Equal(1, finalCount - initialCount);
+
+            using SqlCommand verifyCommand = new SqlCommand($"SELECT str_col FROM {destinationTable}", destinationConnection);
+            using SqlDataReader verifyReader = verifyCommand.ExecuteReader(CommandBehavior.SequentialAccess);
+
+            byte[] expectedBytes = Encoding.UTF8.GetBytes("test");
+
+            Assert.True(verifyReader.Read(), "No data found in destination table after bulk copy.");
+
+            byte[] actualBytes = Encoding.UTF8.GetBytes(verifyReader.GetString(0));
```

Comment:
Instead of getting the string, can you please get the str_col as Byte array? 
The intention is to avoid going through the string conversions/processing of .Net and look at the binary that the server sees. 
This would be a better way of testing, since this is how the bug was found in the first place. 

Essentially the query in SqlCommand would change to `select cast(str_col as varbinary) from {destinationTable};` 

Get the bytes array using GetObject on the reader and then do the binary comparison. 

Let me know if you think otherwise.

Reply:
That is what my original thought was too. Just out of curiosity, I decided to check if the retrieved string contained BOM when data was retrieved with `GetString()` and then converted it into bytes and I could see BOM bytes being present in the byte[] array. 
However, I agree, converting it to `varbinary` leaves no room for any confusion.